### ISSUE TITLE
Conditional Selections

### ISSUE THREAD

**Issue by gksander** [FormidableLabs/groqd#13]
We'd love to support "conditional selections" a la groq's `=>` operator. Ideally, we'd have something that looks like this:

```ts
q(
	q.grab({
		_id: q.string(),
		name: q.string(),
		`name == "Charmander" =>`: {
			name: q.literal("Charmander"),
			hp: ["base.HP", q.number()]
		},
		`base.HP >= 40 =>` {
			isStrong: ["true", q.literal(true)]
		},
		`base.Defense >= 50 =>` {
			isSturdy: ["true", q.literal(true)]
		}
	})
)
```

(This is a simplified example, more realistic examples would likely involved arrays of references to different document types and wanting to select different fields from different document types.)

**Comment by gksander** [comment]
In the simple example above, I _think_ the query's return type should look something like:

```ts
type Base = { _id: string; name: string };  
type Charmander = { name: "Charmander"; hp: number };  
type Strong = { isStrong: true; };

type QueryResult = Base | (Base & Charmander) | (Base & Strong) | (Base & Charmander & Strong);
```

The problem with this is that the more conditional selections you add, the more egregious this `QueryResult` union becomes. If there are $n$ conditional selections, then the following function $f$ models how many constituent parts there will be in the type union:

$$
f(n) = \sum_{i = 0}^{n} {n \choose i}
$$

For 2 conditional selections like above, we have $f(2) = 4$, but if you end up with say 8 conditional selections then you end up with $f(8) = 256$ constituent parts and this quickly blows out of control.

If we were to assume that each conditional selection is mutually exclusive (e.g., you expect to match one conditional selection _at most_, which is a somewhat common scenario if you're selecting fields based on a field like `_type`), then this situation becomes a fair bit easier to deal with because the union size is effectively $n + 1$ in size.

**Comment by nonphoto** [comment]
I just gave this a try today. Unfortunately I don't have a reproduction right now that I can share, but I noticed for a query like this:
```
q( ..., 
    "sections": q("sections", q.filter(), q.grab(
      {},
      {
        "_type == 'menuListSection'": menuListSectionProjection,
        "_type == 'menuComboSection'": menuComboSectionProjection,
      }
    )
)
```
the result from the groq query returned the correct array of `menuListSection`s and `menuComboSection`s, but the using `schema.parse` would coerce the whole array into `[{}, {}, {}, ...]`.

**Comment by gksander** [comment]
Hi @nonphoto – I'll take a look into this to see if I can reproduce!

**Comment by gksander** [comment]
I'm struggling a bit with getting a repro here, is there any chance you can share some more details – even if it's just the full groqd query that you're running, including `menuListSectionProjection` and `menuComboSectionProjection` projection schemas?

**Comment by nonphoto** [comment]
I think this was an issue with my query actually. It was something like this: 
```
{
  "_type == 'menuListSection'": {
    _type: q.literal("menuComboSection"),
    ...
  },
}
```
I would expect this to throw an error on `schema.parse` instead of coercing the result to an empty object though.

**Comment by fritz-c** [comment]
Have you considered making the types less strict, including the conditionally added fields as optional values on a single form of the object instead of separate parts of a union?

e.g.,
```ts
type Result = {
  _id: string;
  name: string;
  hp?: number;
  isStrong?: true;
  isSturdy?: true;
}
```

The conditions for the `isStrong` and `isSturdy` fields in your first example would seem to fit better with such an implementation.

**Comment by gksander** [comment]
Oh man, sorry @fritz-c – I should have mentioned that the "conditional selections" have been implemented via #14, and ended up looking a little bit different than my thought-vomit in the original post.

I need to close this issue soon, but wanted to address the "coercing array to empty objects", since I'm pretty sure I know what that's happening (but haven't had a chance to address it yet)

**Comment by gksander** [comment]
> I just gave this a try today. Unfortunately I don't have a reproduction right now that I can share, but I noticed for a query like this:
> 
> ```
> q( ..., q.grab({
>   "sections": q("sections", q.filter(), q.grab(
>     {},
>     {
>       "_type == 'menuListSection'": menuListSectionProjection,
>       "_type == 'menuComboSection'": menuComboSectionProjection,
>     }
>   )
> })
> ```
> 
> the result from the groq query returned the correct array of `menuListSection`s and `menuComboSection`s, but the using `schema.parse` would coerce the whole array into `[{}, {}, {}, ...]`.

@nonphoto Thank you for pointing this out! It sounds like you found a solution to your problem, but I _think_ the reason your response was getting coerced into an array of empty objects is because the base selection (first arg to `grab`) is an empty object and so `groqd` is generating an empty object schema that's being used. I believe that _is_ a bug in the library, and if no fields are passed to the base selection, we should not include that empty object schema. I'll file an issue for this, should be pretty easy to patch up – and I doubt you'll be the last one to stumble on this exact issue.
